lib/grandpa/message_handler.go

```diff
@@ -427,7 +427,13 @@ func (s *Service) VerifyBlockJustification(hash common.Hash, justification []byt
 	}
 
 	if has {
-		return nil, fmt.Errorf("already have finalised block with setID=%d and round=%d", setID, fj.Round)
+		storedFinalisedHash, err := s.blockState.GetFinalisedHash(fj.Round, setID)
+		if err != nil {
+			return nil, fmt.Errorf("getting finalised hash: %w", err)
+		}
+		if storedFinalisedHash != hash {
+			return nil, fmt.Errorf("%w, setID=%d and round=%d", errFinalisedBlocksMismatch, setID, fj.Round)
+		}
```

Comment:
do we really want to continue scale encoding the justification and then calling `s.finalityGadget.VerifyBlockJustification` in `chainProcessor.handleJustification` if no error is returned? I'm assuming this has already been done in a previous call to `Service.VerifyBlockJustification`

Reply:
> do we really want to continue scale encoding the justification and then calling s.finalityGadget.VerifyBlockJustification

justification is already scale encoded at this point. I don't see a scale.Marshal before this point, so we already got it marshaled. 

I see that VerifyBlockJustification returns justification after scale encoding it again, which is not necessary. I have removed returning justification there. Is that what you meant?

Reply:
Let me put it this way.  Say we enter this `has == true` case so we have a finalised block for a given round and set id. and the `storedFinalisedHash == hash`.  Based on the current code we will exit this `if` block and continue with lines 439 to line 537 where we `s.blockState.SetFinalisedHash(hash, fj.Round, setID)`.  Is this really required given that we found the block already?